url0c7746c72b_wsuetholz_DiscountStrategy_tgz-pJ8-net_suetholz_pos_MainJ8/src/net/suetholz/pos/storage/FakePerstantStorage.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package net.suetholz.pos.storage;
+
+import net.suetholz.pos.Product;
+import net.suetholz.pos.api.CustomerStrategy;
+import net.suetholz.pos.api.DiscountStrategy;
+import net.suetholz.pos.api.PersistantStorageStrategy;
+import net.suetholz.pos.customer.BasicCustomer;
+import net.suetholz.pos.customer.CorporateCustomer;
+import net.suetholz.pos.customer.PreferredCustomer;
+import net.suetholz.pos.discounts.DiscountByFlatRate;
+import net.suetholz.pos.discounts.DiscountByPercentage;
+import net.suetholz.pos.discounts.DiscountByQuantity;
+import net.suetholz.pos.discounts.NoDiscount;
+
+/**
+ *
+ * @author wsuetholz
+ */
+public class FakePerstantStorage implements PersistantStorageStrategy {
+
+    private static final int ARRAY_INCREMENT = 10;	// If array needs to grow, grow by this amount...
+
+    private FakePersistantCustomer[] customers;
+    private int numCustomers;
+    private FakePersistantProduct[] products;
+    private int numProducts;
+    private FakePersistantDiscount[] discounts;
+    private int numDiscounts;
+
+    public FakePerstantStorage(FakePersistantCustomer[] customers, FakePersistantProduct[] products, FakePersistantDiscount[] discounts) {
+	if (customers == null || products == null) {
+	    throw new IllegalArgumentException();
+	}
+
+	this.customers = customers;
+	this.numCustomers = this.customers.length;
+	this.products = products;
+	this.numProducts = this.products.length;
+	this.discounts = discounts;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;discounts&#39; (index: 2 | type: FakePersistantDiscount[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;products&#39;, &#39;customers&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).